---
title: "*Enterococcus* diversity and BSI risk"
engine: knitr
format:
  html:
    code-fold: true
    code-summary: "Show the code"
---

We are now going to expand the exploration of *Enterococcus* diversity within patients with *E. faecium* domination to include all the metagenomes available at MSKCC, including thousands that are not public, and find preliminary links to BSI risk. I have the following goals:

- Quantify the distribution of patient samples with dominations and BSIs that have available, unpublished, metagenomic data.

- Assemble the metagenomes and obtain high quality *E. faecium* MAGs from all those samples.

- Quantify strain-level diversity.
    - Phylogeny of HQ MAGs
    - ST assignments
    - Strain sharing quantification

- Test for potential lineages/STs/Strains enriched in patients with BSI

## Distribution of patient samples with dominations in extended metagenomic data

First, we retrieve the table with all the `lilac` paths to the raw fastqs as a proxy for samples with available megagenomic data:

```{r}
# Get the table with all metagenomic libraries available
vdbR::connect_database()
vdbR::get_table_from_database("rawfastq")

# Add column with Sample ids
rawfastq <- rawfastq %>%
    mutate(Sample = str_extract(path, "(?<=Sample_)(.*?)(?=_IGO)"))
```

Let's now get a list of all the samples for which there are metagenomes available in Isabl, including samples form patients that were dominated or infected by *E. faecium*. Keep in mind that all but one of the cases where BSI was not preceded by domination in the data are actually cases where we don't have microbiome data in the days preceding the infection, but they likely had domination. 

Here, I had to cross-reference the SD paper data with a table from Angel because some of the sample nomenclature was different between the paper, the isabl record, and the `rawfastq` table:

```{r}
# Load required libraries and functions
suppressMessages(library(tidyverse))
source("code/rfunctions/data_helpers.r")

# Prepare genus-level counts
tblrel_genus_meta <- prep_taxa_counts("genus")

# Summarize patients with Enterococcus dominations and infections
summary_data <- tblrel_genus_meta %>%
  group_by(PatientID) %>%
  summarize(
    ever_dominated = any(Enterococcus_abund >= 0.10, na.rm = TRUE),
    ever_infected = any(Enterococcus_infection == TRUE, na.rm = TRUE)
  )

# Identify PatientIDs who were ever dominated or infected
enterococcus_target_patients <- summary_data %>%
  filter(ever_dominated | ever_infected) %>%
  pull(PatientID)

# Get Sample IDs from tblrel_genus_meta for those patients
enterococcus_target_samples <- tblrel_genus_meta %>%
  filter(PatientID %in% enterococcus_target_patients) %>%
  pull(Sample)

# Load records found by Angel for which I have 16S data and metadata
angel_records <- read_csv("misc_files/enterococcus_diversity/22_samples.csv") %>%
    filter(in_isabl & sampleid %in% enterococcus_target_samples) %>%
    pull(sampleid)

# Samples relevant to Enterococcus with metagenomes, including Angel records, no duplicates
enterococcus_sequenced_samples <- union(
    rawfastq %>%
        filter(Sample %in% enterococcus_target_samples) %>%
        distinct(Sample) %>%
        pull(Sample),
    angel_records
)

# All samples from the SD paper with metagenoms
all_target_samples <- union(
    tblrel_genus_meta %>%
        filter(Sample %in% rawfastq$Sample) %>%
        pull(Sample),
    angel_records
)

# Write all_target_samples to a file as a single line, comma-separated, no quotes
writeLines(paste(all_target_samples, collapse = ","), "misc_files/enterococcus_diversity/all_target_samples.txt")
```

We now get the paths to the host-depleted reads for all these samples:

```{bash, eval=FALSE}
# Activate isabl conda
conda activate /usersoftware/collab004/conda/isablprod

# Set env variables
export ISABL_CLIENT_ID=1
export ISABL_API_URL="https://isabl.microbiome.mskcc.org/api/v1"

# Load sample names to variable
#ALL_TARGET_SAMPLES=$(cat misc_files/enterococcus_diversity/all_target_samples.txt | cut -d ',' -f 1,2)
ALL_TARGET_SAMPLES=$(cat misc_files/enterococcus_diversity/all_target_samples.txt)

# Get table with lilac paths to preprocessed reads
isabl get-metadata analyses \
  -fi application.pk 43 \
  -fi targets.sample.identifier.in ${ALL_TARGET_SAMPLES} \
  --field pk \
  --field targets.sample.identifier \
  --field results.fq1 \
  --field results.fq2 \
  > misc_files/enterococcus_diversity/all_target_samples_paths.txt

# Get list of unique sample identifiers with available host-depleted reads
cat misc_files/enterococcus_diversity/all_target_samples_paths.txt | cut -f 2 | sort | uniq > misc_files/enterococcus_diversity/all_target_samples_hostdepleted.txt
```

Let's check which of the samples do not have available host-depleted reads, which Mirae and Nick will run through the pipeline and I can include them later:

```{r}
# Load list of samples with host-depleted reads
all_samples_hostdepleted <- scan("misc_files/enterococcus_diversity/all_target_samples_hostdepleted.txt", what = "character") %>%
    str_remove_all("")

samples_missing <- setdiff(all_target_samples, all_samples_hostdepleted)
writeLines(samples_missing, "misc_files/enterococcus_diversity/samples_missing_hostdeplet.txt")
```

I also noticed that some of the samples have multiple records of host depletion outputs, so I checked which are duplicated:

```{bash, eval=FALSE}
# Find duplicated sample names in the file
cut -f2 misc_files/enterococcus_diversity/all_target_samples_paths.txt | sort | uniq -d
```

After talking with Mirae, it seems these cases correspond to duplicated outputs from the same library, so I will only take one record from each sample (2,077 unique) for the assembly pipeline.

## Metagenome assembly and binning

### metaSPAdes assembly

I preprocessed the reads with fastp and generated a metagenomic assembly with metaSPAdes for each metagenomic library:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity/metaspades_isabl1.sh
```

Some of the assemblies failed due to memory issues, but I will continue working with the 2022 that did work. This generates the list of samples with succesful assemblies:

```{bash, eval=FALSE}
find analyses/enterococcus_diversity/metagenomes/assembly/metaspades -type f -name contigs.fasta \
  | cut -d'/' -f6 \
  | sort > misc_files/enterococcus_diversity/metaspades_successful_samples.txt
```

### Multi-sample binning with VAMB

I generated the files with the list of samples sequenced with metagenomes for each patient (565): 

```{r, eval=FALSE}
# Load list of successful sample IDs
successful_samples <- read_lines("misc_files/enterococcus_diversity/metaspades_successful_samples.txt")

# Filter tblrel_genus_meta for those samples
patient_samples <- tblrel_genus_meta %>%
  filter(Sample %in% successful_samples) %>%
  select(PatientID, Sample) %>%
  distinct()

# Create directory for patient sample lists
dir.create("analyses/enterococcus_diversity/binning/short/isabl1/patient_samples", recursive = TRUE, showWarnings = FALSE)

# Save SampleIDs for each PatientID to separate .txt files
unique_patient_ids <- unique(patient_samples$PatientID)
for (patient_id in unique_patient_ids) {
  sample_ids <- patient_samples$Sample[patient_samples$PatientID == patient_id]
  file_path <- paste0("analyses/enterococcus_diversity/binning/short/isabl1/patient_samples/", patient_id, ".txt")
  writeLines(sample_ids, file_path)
}
```

These files are needed to arrange assembly concatenation and read mapping. After that, I concatenated the assemblies from each patient, mapped the reads, and ran multi-sample VAMB:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity/vamb_multi_isabl1.sh
```

Then I compiled the bins from VAMB:

```{bash, eval=FALSE}
SRC_BASE="analyses/enterococcus_diversity/binning/short/isabl1/vamb/multi"
DEST_BASE="analyses/enterococcus_diversity_II/binning/short/isabl1/vamb/multi_bins"

mkdir -p "${DEST_BASE}"

find "${SRC_BASE}" -mindepth 1 -maxdepth 1 -type d | while read -r patient_dir; do
    sample=$(basename "${patient_dir}")
    bins_dir="${patient_dir}/vambout/bins"
    if [ -d "${bins_dir}" ]; then
        for fna in "${bins_dir}"/*.fna; do
            [ -e "$fna" ] || continue
            bin_name=$(basename "$fna")
            cp "$fna" "${DEST_BASE}/${sample}_${bin_name}"
        done
    fi
done
```

And I classified the bins using GTDB:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/isabl1_gtdb.sh
```

With the classification results, I compiled the *Enterococcus* MAGs:

```{bash, bash=FALSE}
# Define input and output paths
GTDB_FILES=("analyses/enterococcus_diversity_II/binning/short/isabl1/gtdb/gtdbtk.bac120.summary.tsv")
BIN_DIRS=("analyses/enterococcus_diversity_II/binning/short/isabl1/vamb/multi_bins")
OUTPUT_DIR="analyses/enterococcus_diversity_II/genomes/mags/isabl1"

# Create the output directory if it doesn't exist
mkdir -p ${OUTPUT_DIR}
# Loop through GTDB files
for i in "${!GTDB_FILES[@]}"; do
  gtdb_file="${GTDB_FILES[$i]}"
  bin_dir="${BIN_DIRS[$i]}"
  
  # Extract filenames of bins classified in the family Bacteroidaceae and copy them
  awk -F'\t' 'NR > 1 && $2 ~ /g__Enterococcus/ {print $1}' "${gtdb_file}" | while read -r bin_file; do
    full_path="${bin_dir}/${bin_file}.fna"
    if [[ -f ${full_path} ]]; then
      cp "${full_path}" "${OUTPUT_DIR}/"
    else
      echo "File ${full_path} not found, skipping..."
    fi
  done
done
```

And I got completeness and contamination estimates for those with CheckM2:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/enterococcus_isabl1_checkm2.sh
```

With the classifications and the CheckM2 results, I selected the nearly complete (>95% completeness, <5% contamination) *Enterococcus faecium* mags:

```{r, eval=FALSE}
# Load required library and custom functions
suppressMessages(library(tidyverse))

# Load and merge GTBD results
meta_gtdb <- suppressMessages(read_tsv("analyses/enterococcus_diversity_II/binning/short/isabl1/gtdb/gtdbtk.bac120.summary.tsv"))

# Load CheckM2 results
checkm <- suppressMessages(read_tsv("analyses/enterococcus_diversity_II/genomes/checkm2/isabl1/quality_report.tsv"))

# Compile E. faecium NC MAG data
efaecium_mag_data <- checkm %>%
  mutate(binning_strategy = case_when(
            str_detect(Name, "single") ~ "single",
            str_detect(Name, "multi") ~ "multi"
          ),
          sample = str_split(Name, "_", simplify = TRUE)[, 2]
        ) %>%
  left_join(meta_gtdb, by = c("Name" = "user_genome")) %>%
  mutate(species = str_remove(classification, ".*s__")) %>%
  filter(Completeness >= 95 & Contamination <= 5 & str_detect(species, "Enterococcus_B faecium")) %>%
  group_by(species, sample) %>%
    slice_max(order_by = Completeness, n = 1, with_ties = FALSE) %>%
    ungroup()

# Print list of paths to NC MAGs for pangenome analysis
write(paste0("analyses/enterococcus_diversity_II/genomes/mags/", efaecium_mag_data$Name, ".fna"), "analyses/enterococcus_diversity/genomes/mags/efaecium_nc_isabl1.txt")
```