---
title: "*Enterococcus* diversity and BSI risk II"
engine: knitr
format:
  html:
    code-fold: true
    code-summary: "Show the code"
---

We are now going to expand the exploration of *Enterococcus* diversity within patients with *E. faecium* domination to include all the metagenomes available at MSKCC, including thousands that are not public, and find preliminary links to BSI risk. I have the following goals:

- Quantify the distribution of patient samples with dominations and BSIs that have available, unpublished, metagenomic data.

- Assemble the metagenomes and obtain high quality *E. faecium* MAGs from all those samples.

- Quantify strain-level diversity.
    - Phylogeny of HQ MAGs
    - ST assignments
    - Strain sharing quantification

- Test for potential lineages/STs/Strains enriched in patients with BSI

## Distribution of patient samples with dominations in extended metagenomic data

First, we retrieve the table with all the `lilac` paths to the raw fastqs as a proxy for samples with available megagenomic data:

```{r}
# Get the table with all metagenomic libraries available
vdbR::connect_database()
vdbR::get_table_from_database("rawfastq")

# Add column with Sample ids
rawfastq <- rawfastq %>%
    mutate(Sample = str_extract(path, "(?<=Sample_)(.*?)(?=_IGO)"))
```

Let's now get a list of all the samples for which there are metagenomes available in Isabl, including samples form patients that were dominated or infected by *E. faecium*. Keep in mind that all but one of the cases where BSI was not preceded by domination in the data are actually cases where we don't have microbiome data in the days preceding the infection, but they likely had domination. 

Here, I had to cross-reference the SD paper data with a table from Angel because some of the sample nomenclature was different between the paper, the isabl record, and the `rawfastq` table:

```{r}
# Load required libraries and functions
suppressMessages(library(tidyverse))
source("code/rfunctions/data_helpers.r")

# Prepare genus-level counts
tblrel_genus_meta <- prep_taxa_counts("genus")

# Summarize patients with Enterococcus dominations and infections
summary_data <- tblrel_genus_meta %>%
  group_by(PatientID) %>%
  summarize(
    ever_dominated = any(Enterococcus_abund >= 0.10, na.rm = TRUE),
    ever_infected = any(Enterococcus_infection == TRUE, na.rm = TRUE)
  )

# Identify PatientIDs who were ever dominated or infected
enterococcus_target_patients <- summary_data %>%
  filter(ever_dominated | ever_infected) %>%
  pull(PatientID)

# Get Sample IDs from tblrel_genus_meta for those patients
enterococcus_target_samples <- tblrel_genus_meta %>%
  filter(PatientID %in% enterococcus_target_patients) %>%
  pull(Sample)

# Load records found by Angel for which I have 16S data and metadata
angel_records <- read_csv("misc_files/enterococcus_diversity/22_samples.csv") %>%
    filter(in_isabl & sampleid %in% enterococcus_target_samples) %>%
    pull(sampleid)

# Samples relevant to Enterococcus with metagenomes, including Angel records, no duplicates
enterococcus_sequenced_samples <- union(
    rawfastq %>%
        filter(Sample %in% enterococcus_target_samples) %>%
        distinct(Sample) %>%
        pull(Sample),
    angel_records
)

# All samples from the SD paper with metagenoms
all_target_samples <- union(
    tblrel_genus_meta %>%
        filter(Sample %in% rawfastq$Sample) %>%
        pull(Sample),
    angel_records
)

# Write all_target_samples to a file as a single line, comma-separated, no quotes
writeLines(paste(all_target_samples, collapse = ","), "misc_files/enterococcus_diversity/all_target_samples.txt")
```

We now get the paths to the host-depleted reads for all these samples:

```{bash, eval=FALSE}
# Activate isabl conda
conda activate /usersoftware/collab004/conda/isablprod

# Set env variables
export ISABL_CLIENT_ID=1
export ISABL_API_URL="https://isabl.microbiome.mskcc.org/api/v1"

# Load sample names to variable
#ALL_TARGET_SAMPLES=$(cat misc_files/enterococcus_diversity/all_target_samples.txt | cut -d ',' -f 1,2)
ALL_TARGET_SAMPLES=$(cat misc_files/enterococcus_diversity/all_target_samples.txt)

# Get table with lilac paths to preprocessed reads
isabl get-metadata analyses \
  -fi application.pk 43 \
  -fi targets.sample.identifier.in ${ALL_TARGET_SAMPLES} \
  --field pk \
  --field targets.sample.identifier \
  --field results.fq1 \
  --field results.fq2 \
  > misc_files/enterococcus_diversity/all_target_samples_paths.txt

# Get list of unique sample identifiers with available host-depleted reads
cat misc_files/enterococcus_diversity/all_target_samples_paths.txt | cut -f 2 | sort | uniq > misc_files/enterococcus_diversity/all_target_samples_hostdepleted.txt
```

Let's check which of the samples do not have available host-depleted reads, which Mirae and Nick will run through the pipeline and I can include them later:

```{r}
# Load list of samples with host-depleted reads
all_samples_hostdepleted <- scan("misc_files/enterococcus_diversity/all_target_samples_hostdepleted.txt", what = "character") %>%
    str_remove_all("")

samples_missing <- setdiff(all_target_samples, all_samples_hostdepleted)
writeLines(samples_missing, "misc_files/enterococcus_diversity/samples_missing_hostdeplet.txt")
```

I also noticed that some of the samples have multiple records of host depletion outputs, so I checked which are duplicated:

```{bash, eval=FALSE}
# Find duplicated sample names in the file
cut -f2 misc_files/enterococcus_diversity/all_target_samples_paths.txt | sort | uniq -d
```

After talking with Mirae, it seems these cases correspond to duplicated outputs from the same library, so I will only take one record from each sample (2,077 unique) for the assembly pipeline.

## Metagenome assembly and binning

### metaSPAdes assembly

I preprocessed the reads with fastp and generated a metagenomic assembly with metaSPAdes for each metagenomic library:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity/metaspades_isabl1.sh
```

Some of the assemblies failed due to memory issues, but I will continue working with the 2022 that did work. This generates the list of samples with succesful assemblies:

```{bash, eval=FALSE}
find analyses/enterococcus_diversity/metagenomes/assembly/metaspades -type f -name contigs.fasta \
  | cut -d'/' -f6 \
  | sort > misc_files/enterococcus_diversity/metaspades_successful_samples.txt
```

### Multi-sample binning with VAMB

I generated the files with the list of samples sequenced with metagenomes for each patient (565): 

```{r, eval=FALSE}
# Load list of successful sample IDs
successful_samples <- read_lines("misc_files/enterococcus_diversity/metaspades_successful_samples.txt")

# Filter tblrel_genus_meta for those samples
patient_samples <- tblrel_genus_meta %>%
  filter(Sample %in% successful_samples) %>%
  select(PatientID, Sample) %>%
  distinct()

# Create directory for patient sample lists
dir.create("analyses/enterococcus_diversity/binning/short/isabl1/patient_samples", recursive = TRUE, showWarnings = FALSE)

# Save SampleIDs for each PatientID to separate .txt files
unique_patient_ids <- unique(patient_samples$PatientID)
for (patient_id in unique_patient_ids) {
  sample_ids <- patient_samples$Sample[patient_samples$PatientID == patient_id]
  file_path <- paste0("analyses/enterococcus_diversity/binning/short/isabl1/patient_samples/", patient_id, ".txt")
  writeLines(sample_ids, file_path)
}
```

These files are needed to arrange assembly concatenation and read mapping. After that, I concatenated the assemblies from each patient, mapped the reads, and ran multi-sample VAMB:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity/vamb_multi_isabl1.sh
```

Then I compiled the bins from VAMB:

```{bash, eval=FALSE}
SRC_BASE="analyses/enterococcus_diversity/binning/short/isabl1/vamb/multi"
DEST_BASE="analyses/enterococcus_diversity_II/binning/short/isabl1/vamb/multi_bins"

mkdir -p "${DEST_BASE}"

find "${SRC_BASE}" -mindepth 1 -maxdepth 1 -type d | while read -r patient_dir; do
    sample=$(basename "${patient_dir}")
    bins_dir="${patient_dir}/vambout/bins"
    if [ -d "${bins_dir}" ]; then
        for fna in "${bins_dir}"/*.fna; do
            [ -e "$fna" ] || continue
            bin_name=$(basename "$fna")
            cp "$fna" "${DEST_BASE}/${sample}_${bin_name}"
        done
    fi
done
```

And I classified the bins using GTDB:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/isabl1_gtdb.sh
```

With the classification results, I compiled the *Enterococcus* MAGs:

```{bash, bash=FALSE}
# Define input and output paths
GTDB_FILES=("analyses/enterococcus_diversity_II/binning/short/isabl1/gtdb/gtdbtk.bac120.summary.tsv")
BIN_DIRS=("analyses/enterococcus_diversity_II/binning/short/isabl1/vamb/multi_bins")
OUTPUT_DIR="analyses/enterococcus_diversity_II/genomes/mags/isabl1"

# Create the output directory if it doesn't exist
mkdir -p ${OUTPUT_DIR}
# Loop through GTDB files
for i in "${!GTDB_FILES[@]}"; do
  gtdb_file="${GTDB_FILES[$i]}"
  bin_dir="${BIN_DIRS[$i]}"
  
  # Extract filenames of bins classified in the family Bacteroidaceae and copy them
  awk -F'\t' 'NR > 1 && $2 ~ /g__Enterococcus/ {print $1}' "${gtdb_file}" | while read -r bin_file; do
    full_path="${bin_dir}/${bin_file}.fna"
    if [[ -f ${full_path} ]]; then
      cp "${full_path}" "${OUTPUT_DIR}/"
    else
      echo "File ${full_path} not found, skipping..."
    fi
  done
done
```

And I got completeness and contamination estimates for those with CheckM2:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/enterococcus_isabl1_checkm2.sh
```

With the classifications and the CheckM2 results, I selected the 671 nearly complete (>95% completeness, <5% contamination) *Enterococcus faecium* mags:

```{r, eval=FALSE}
# Load required library and custom functions
suppressMessages(library(tidyverse))

# Load and merge GTBD results
meta_gtdb <- suppressMessages(read_tsv("analyses/enterococcus_diversity_II/binning/short/isabl1/gtdb/gtdbtk.bac120.summary.tsv"))

# Load CheckM2 results
checkm <- suppressMessages(read_tsv("analyses/enterococcus_diversity_II/genomes/checkm2/isabl1/quality_report.tsv"))

# Compile E. faecium NC MAG data
efaecium_mag_data <- checkm %>%
  mutate(binning_strategy = case_when(
            str_detect(Name, "single") ~ "single",
            str_detect(Name, "multi") ~ "multi"
          ),
          sample = str_split(Name, "_", simplify = TRUE)[, 2]
        ) %>%
  left_join(meta_gtdb, by = c("Name" = "user_genome")) %>%
  mutate(species = str_remove(classification, ".*s__")) %>%
  filter(Completeness >= 95 & Contamination <= 5 & str_detect(species, "Enterococcus_B faecium")) %>%
  group_by(species, sample) %>%
    slice_max(order_by = Completeness, n = 1, with_ties = FALSE) %>%
    ungroup()

# Print list of paths to NC MAGs for pangenome analysis
write(paste0("analyses/enterococcus_diversity_II/genomes/mags/isabl1/", efaecium_mag_data$Name, ".fna"), "analyses/enterococcus_diversity_II/genomes/mags/efaecium_nc_isabl1.txt")
```

## *E. faecium* phylogeny and ST inference

### Phylogeny with Lebreton et al 2013 reference

The phylogeny will include the isabl1 NC mags and the genomes from Lebreton et al as a reference. Let's start by getting a list of all genomes included:

```{bash, eval=FALSE}
# Paths
NC_MAGS="analyses/enterococcus_diversity_II/genomes/mags/efaecium_nc_isabl1.txt"
PHYLO_DIR="analyses/enterococcus_diversity_II/phylogenetics/isabl1_plus_lebreton"
LEBRETON_DIR="analyses/enterococcus_diversity/genomes/lebreton_2013"
LEBRETON_FASTA_DIR="analyses/enterococcus_diversity/genomes/lebreton_2013/fasta"

# File with paths to Lebreton genomes
ls -d $LEBRETON_FASTA_DIR/*.fna > ${LEBRETON_DIR}/genome_paths.txt

# Driectory for phylogenetic analyses
mkdir -p ${PHYLO_DIR}

# List of genomes included in phylogenetic analyes
cat ${NC_MAGS} ${LEBRETON_DIR}/genome_paths.txt > ${PHYLO_DIR}/genome_list.txt
```

And then I ran ggCaller to get a core genome alignment and an NJ tree:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/ggcaller_align_isabl1.sh
rm analyses/enterococcus_diversity_II/genomes/mags/isabl1/*.fmp
```

### ST inference

I classified the MAGs into the STs scheme defined by [Homan et al. 2002](https://journals.asm.org/doi/10.1128/jcm.40.6.1963-1971.2002) using the `mlst` software:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/mlst_isabl1.sh
```

### Visualizing *E. faecium* diversity and BSI incidence

Let's first compile all the metadata for these MAGs

```{r, eval=FALSE}
# Load and merge GTBD results
meta_gtdb <- suppressMessages(read_tsv("analyses/enterococcus_diversity_II/binning/short/isabl1/gtdb/gtdbtk.bac120.summary.tsv"))

# Load CheckM2 results
checkm <- suppressMessages(read_tsv("analyses/enterococcus_diversity_II/genomes/checkm2/isabl1/quality_report.tsv"))

# Load MLST classifications
mlst_2002 <- read.csv("analyses/enterococcus_diversity_II/genomes/mlst/mlst_2002_isabl1.csv", header = FALSE, fill = TRUE)
colnames(mlst_2002) <- c("genome_path",
                          "scheme",
                          "ST",
                          paste0("allele_", seq(1:7))
                          )
mlst_2002 <- mlst_2002 %>%
  mutate(genome = tools::file_path_sans_ext(basename(genome_path)))


# Compile MAG data
enterococcus_mag_data <- checkm %>%
  left_join(meta_gtdb, by = c("Name" = "user_genome")) %>%
  left_join(mlst_2002, by = c("Name" = "genome")) %>%
  mutate(species = str_remove(classification, ".*s__")) %>%
  filter(Completeness >= 70 & Contamination <= 5 & str_detect(classification, "g__Enterococcus"))

# Load Lebreton metadata
lebreton_metadata <- suppressMessages(
  read_csv("data/lebreton_et_al_2013_genomes.csv") %>%
    mutate(tiplabel = str_remove(Accession, "\\.1"))
)
```

And now we plot different versions of the NJ tree to highlight:

1. Main clades and and distribution of hospital-associated *E. faecium. Note that in this version, there are many MSK strains in clade A2. Does

```{r}
# Load phylogenetics packages
suppressMessages(library(ape))
suppressMessages(library(ggtree))
suppressMessages(library(ggtreeExtra))
suppressMessages(library(treeio))

# Load the NJ tree
lebreton_plus_mags_tree <- read.tree("analyses/enterococcus_diversity_II/phylogenetics/isabl1_plus_lebreton/ggcaller/core_tree_NJ.nwk")

# Remove single quotation marks from tip labels
lebreton_plus_mags_tree$tip.label <- gsub("'", "", lebreton_plus_mags_tree$tip.label)

# Root the tree with Clade B
outgroup <- lebreton_metadata %>% filter(Clade == "B") %>% pull(tiplabel)
lebreton_plus_mags_tree <- root(lebreton_plus_mags_tree, outgroup = outgroup, resolve.root = TRUE)

# MAG source metadata
mag_source_data <- enterococcus_mag_data %>%
  filter(Name %in% lebreton_plus_mags_tree$tip.label) %>%
  mutate(tiplabel = Name,
    source = "MSK") %>%
  select(tiplabel, source)

# Lebreton source metadata
hospital_categories <- c("Hosp_Feces", "Clinical Isolate", "Hospital Surveillance", "Hospital Outbreak", "Hospital Unknown")
lebreton_source_data <- lebreton_metadata %>%
  mutate(source = case_when(Category %in% hospital_categories ~ "Hospital",
                            Category == "Non-hospitalized individual" ~ "Human commensal",
                            .default = Category)) %>%
  select(tiplabel, source)

# Compile source metadata and add infection data
source_data <- bind_rows(mag_source_data, lebreton_source_data) %>%
  mutate(
    PatientID = case_when(
      source != "MSK" ~ NA_character_,
      str_detect(tiplabel, "pt") ~ stringr::word(tiplabel, 1, -4, sep = "_"),
      TRUE ~ stringr::word(tiplabel, 1, sep = "_")
    )
  ) %>%
  left_join(summary_data, by = "PatientID")

clade_a1_node <- MRCA(lebreton_plus_mags_tree, c("GCF_000322425", "GCF_000239095"))

# Merge tree tip labels with source data
tree_data_main <- as_tibble(lebreton_plus_mags_tree) %>%
  left_join(source_data, by = c("label" = "tiplabel"))

# Define colors and shapes for sources
source_colors <- c(
  "MSK" = "#f4a261",          # Warm orange
  "Hospital" = "#e63946",     # Deep red
  "Human commensal" = "#2a9d8f", # Teal green
  "Animal" = "#89c2d9",       # Light blue
  "Other" = "gray50"          # Neutral gray
)
source_shapes <- c("MSK" = 16, "Hospital" = 16, "Human commensal" = 16, "Animal" = 16, "Other" = 16)

# Plot the tree with tip shapes and colors
ggtree(lebreton_plus_mags_tree, layout = "circular") %<+% tree_data_main +
  geom_tippoint(aes(color = source, shape = source), size = 3) +
  scale_color_manual(values = source_colors) +
  scale_shape_manual(values = source_shapes) +
  geom_hilight(node = clade_a1_node, fill = "gray75", alpha = 0.4, to.bottom = T) +
  geom_cladelab(node = clade_a1_node, label = "A1", barsize = 0, fontsize = 8, offset.text = 6000) +
  geom_strip(taxa1 = "GCF_000322285", taxa2 = "GCF_000321765",
              barsize = 2,
              label = "A2",
              offset = 1000,
              offset.text = 3000,
              fontsize = 8) +
  geom_strip(taxa1 = "GCF_000157655", taxa2 = "GCF_000322145",
              barsize = 2,
              label = "B",
              offset = 4000,
              offset.text = 3000,
              fontsize = 8,
              align = F) +
  theme(legend.position = "right")

# Save the plot
#ggsave("document/plots/lebreton_plus_mags_main.pdf")
```

2. Clade A1 highlighting the STs:

```{r}
# Clade A1 taxa
clade_a1_tips <- c(filter(lebreton_metadata, Clade == "A1") %>% pull(tiplabel), pull(mag_source_data, tiplabel))

# Prune the tree to keep only clade A1 tips
clade_a1_tree <- keep.tip(lebreton_plus_mags_tree, clade_a1_tips)

# ST data for taxa in Clade A1
st_data <- source_data %>%
  left_join(select(enterococcus_mag_data, Name, ST), by = c("tiplabel" = "Name"))

# Join tree with data
tree_data_st <- as_tibble(clade_a1_tree) %>%
  left_join(st_data, by = c("label" = "tiplabel"))

# Plot the tree with tip labels colored by ST to check 
ggtree(clade_a1_tree, layout = "rectangular") %<+% tree_data_st +
  geom_tippoint(aes(color = ST), size = 3) +
  labs(color = "ST") +
  theme(legend.position = "right") +
  ggnewscale::new_scale("shape") +
  geom_fruit(
    geom = geom_point,
    mapping = aes(y = label, shape = as.character(ever_infected)),
    size = 3,
    color = "black",
    pwidth = 0.05,
    axis.params = list(axis = "y", title = "Ever infected")
  ) +
  scale_shape_manual(
    name = "Ever infected",
    values = infected_shapes,
    labels = c("No", "Yes")
  )
```

Now, I plotted the distribution of infected and uninfected patients across STs:

```{r}
# Summarize: count unique PatientIDs per ST and ever_infected status
st_patient_counts <- st_data %>%
  filter(!is.na(PatientID), !is.na(ST)) %>%
  group_by(ST, ever_infected, PatientID) %>%
  summarise(.groups = "drop") %>%  # one row per ST, ever_infected, PatientID
  group_by(ST, ever_infected) %>%
  summarise(n_patients = n_distinct(PatientID), .groups = "drop")

# Calculate total patients per ST for ordering
st_totals <- st_patient_counts %>%
  group_by(ST) %>%
  summarise(total_patients = sum(n_patients), .groups = "drop")

# Set ST as a factor ordered by total_patients
st_patient_counts <- st_patient_counts %>%
  left_join(st_totals, by = "ST") %>%
  mutate(ST = factor(ST, levels = st_totals$ST[order(st_totals$total_patients, decreasing = TRUE)])) %>%
  filter(ST != "-")

# Plot colonization and infections by ST
ggplot(st_patient_counts, aes(x = ST, y = n_patients, fill = ever_infected)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = c("TRUE" = "#0D7E2B", "FALSE" = "gray75"),
                    name = "Infection status",
                    labels = c("Not infected", "Infected")) +
  labs(x = "ST", y = "Number of patients") +
  custom_theme +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = c(0.75, 0.8)
  )

# Save the plot
ggsave("document/plots/st_bsi_distribution_isabl1.pdf", width = 3, height = 4)
```

### Strain sharing in patients colonized with ST412

I selected ST412 because it is the most common strain in this segment of the dataset. Let's first get a list of the samples with ST412:

```{r}
# ST412 samples
st412_samples <- st_data %>%
  filter(ST == "412") %>%
  mutate(sample = str_remove(tiplabel, paste0(PatientID, "_"))) %>%
  mutate(sample = str_remove(sample, "_C_.*")) %>%
  pull(sample)

dir.create("analyses/enterococcus_diversity_II/strains/instrain/isabl1/st412/misc", recursive = TRUE, showWarnings = FALSE)

# Write ST412 samples
write(st412_samples, "analyses/enterococcus_diversity_II/strains/instrain/isabl1/st412/misc/st42_samples.txt")
```

With these samples, I can now infer the inStrain profiles for the ST412 genomes using the dereplicated library of MAGs and the reference *E. faecium* MAG I used for the previous iteration of enterococcus diversity analyses ADD LINK HERE:

```{bash, eval=FALSE}
# Submit the first inStrain profile job (for the first sample) and capture its job ID
jid=$(sbatch --array=1-1 code/scripts/enterococcus_diversity_II/instrain_profile_isabl1_st412.sh | awk '{print $4}')

# Submit the rest of the inStrain profile jobs (samples 1-172), dependent on the first finishing
sbatch --dependency=afterok:$jid --array=2-172 code/scripts/enterococcus_diversity_II/instrain_profile_isabl1_st412.sh
```

And then, I compared the profiles:

```{bash, eval=FALSE}
sbatch code/scripts/enterococcus_diversity_II/instrain_compare_st42.sh
```

I wanted to visualize the relationship between strain sharing (popANI) and the time between stool sample collection from different patients to determine if strain sharing is more common when samples were taken closely in time. For this, I first needed to ggenerate a table of the time bewteen sample collection for the stool samples that were shotgun sequenced:

```{r}
# Load the sample metadata
meta_tblASVsamples <- suppressMessages(
  read_csv("data/tblASVsamples.csv",
    col_types = cols(
      .default = col_guess(),
      Pool = col_character()
    )) %>%
  #filter(!is.na(AccessionShotgun)) %>%
  #mutate(SampleID = paste0("s_", SampleID)) %>%
  distinct(SampleID, .keep_all = TRUE)
)

# Load the patient dictionary
patient_dictionary <- suppressMessages(
  read_csv("data/tCastoriPatientDictionaryWithDatesAndOtherStudiesPatientId.txt")
)

# Define column types for dictionary
patient_dictionary <- patient_dictionary %>%
  mutate(
    cmicId = as.character(cmicId),
    referenceDate = as.Date(referenceDate, format = "%Y-%m-%d")
  )

# Merge the sample metadata with the patient dictionary to get reference dates
meta_tblASVsamples_with_dates <- meta_tblASVsamples %>%
    left_join(patient_dictionary, by = c("PatientID" = "cmicId")) %>%
    mutate(
        collectionDate = referenceDate + days(Timepoint) # Convert patient-specific timepoints to actual dates
    )

# Generate all pairwise combinations of samples from different patients
time_difference <- meta_tblASVsamples_with_dates %>%
    select(SampleID, PatientID, collectionDate) %>%
    rename(sample1 = SampleID, patient1 = PatientID, date1 = collectionDate) %>%
    full_join(
        meta_tblASVsamples_with_dates %>%
            select(SampleID, PatientID, collectionDate) %>%
            rename(sample2 = SampleID, patient2 = PatientID, date2 = collectionDate),
        by = character()
    ) %>%
    mutate(
        time_between_sample_collection = abs(as.numeric(difftime(date1, date2, units = "days")))
    ) %>%
    select(sample1, sample2, time_between_sample_collection, patient1, patient2)

time_difference
```



```{r}
# Load inStrain comparison table
instrain_compare <- read_tsv("analyses/enterococcus_diversity_II/strains/instrain/isabl1/st412/instrain_compare/output/instrain_compare_genomeWide_compare.tsv") %>%
  mutate(sample1 = str_remove(name1, ".sorted.bam"),
        sample2 = str_remove(name2, ".sorted.bam"),
        popANI = popANI*100) %>%
  left_join(time_difference, by = c("sample1", "sample2")) %>%
  mutate(patient1_ord = if_else(patient1 < patient2, patient1, patient2),
        patient2_ord = if_else(patient1 < patient2, patient2, patient1)) %>%
  left_join(summary_data, by = c("patient1" = "PatientID")) %>%
  left_join(summary_data, by = c("patient2" = "PatientID")) %>%
  rename(patient1_infection = ever_infected.x,
        patient2_infection = ever_infected.y) %>%
  mutate(pair = case_when(
                  patient1_infection & patient2_infection ~ "both",
                  !patient1_infection & !patient2_infection ~ "none",
                  !patient1_infection & patient2_infection | patient1_infection & !patient2_infection ~ "one"
  )) %>%
  filter(percent_compared >= 0.85)

# Identify patients with NA values in instrain_compare
na_patients <- instrain_compare %>%
  filter(is.na(patient1) | is.na(patient2)) %>%
  select(sample1, sample2, patient1, patient2) %>%
  distinct()

# Comparisons between samples from the same patients
instrain_compare_within <- instrain_compare %>%
  filter(patient1_ord == patient2_ord) %>%
  mutate(log10_popANI = log10(100 - popANI))

# Comparisons between samples from different patients
instrain_compare_between <- instrain_compare %>%
  filter(patient1_ord != patient2_ord) %>%
  mutate(log10_popANI = log10(100 - popANI))

# Comparisons with the highest popANI for each pair of patients
instrain_compare_max_popANI <- instrain_compare_between %>%
  group_by(patient1_ord, patient2_ord) %>%
  slice_max(order_by = popANI, n = 1, with_ties = FALSE) %>%
  ungroup() %>%
  # Remove any comparisons where either patient is 1277 or 1738
  filter(!(patient1 %in% c("1277", "1728") | patient2 %in% c("1277", "1728"))) # Removing mislabe, fix later

# Plot with the highest popANI for each pair of patients 
between_patient_max_plot <- ggplot(instrain_compare_max_popANI, aes(x = time_between_sample_collection, y = log10(100 - popANI), fill = pair)) +
  geom_point(alpha = 0.9, shape = 21, color = "black", size = 2.5) +
  geom_hline(yintercept = log10(100 - 99.999), color = "red", linetype = "dashed") +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 6)) +
  scale_y_reverse(
    breaks = seq(-4, -1, by = 1),
    labels = c("1", "10", "100", "1000"),
    limits = c(-1, -4.4)
  ) +
  scale_fill_manual(
    values = c("none" = "#D1E5F0", "one" = "#B2182B", "both" = "#2166AC"),
    labels = c("none" = "not infected vs not infected",
               "one"  = "infected vs not infected",
               "both" = "Infected vs infected")
  ) +
  labs(
    x = "Time between patient sample collection (days)",
    y = "Nucleotide differences per million bp",
    fill = "Patient comparison:"
  ) +
  custom_theme +
  theme(legend.position = c(0.71, 0.82))

# Print the plots
between_patient_max_plot

# Save the plot
ggsave("document/plots/strain_sharing_isabl1.pdf",
  plot = between_patient_max_plot, width = 5, height = 4)
```


## Trash
```{r}
# Use your table
tbl <- instrain_compare_max_popANI

# Ensure patient columns are character
tbl <- tbl %>% mutate(patient1 = as.character(patient1), patient2 = as.character(patient2))

# 1) Mirror rows so we have both (a,b) and (b,a)
tbl_sym <- tbl %>%
  select(patient1, patient2, popANI, patient1_infection, patient2_infection) %>%
  mutate(popANI = as.numeric(popANI)) %>%
  bind_rows(tbl %>% select(patient1 = patient2, patient2 = patient1, popANI, patient1_infection = patient2_infection, patient2_infection = patient1_infection))

# 2) Add diagonal entries (patient vs same patient) with popANI = 100 and infection flags
patients <- sort(unique(c(tbl_sym$patient1, tbl_sym$patient2)))
diag_df <- tibble(
  patient1 = patients,
  patient2 = patients,
  popANI = 100,
  patient1_infection = NA,
  patient2_infection = NA
)

tbl_sym <- tbl_sym %>%
  bind_rows(diag_df) %>%
  distinct(patient1, patient2, .keep_all = TRUE)

# 3) Make sure we have a single infection flag per patient:
# prefer non-NA values from patient1_infection/patient2_infection columns
infect1 <- tbl %>% select(patient = patient1, infected = patient1_infection)
infect2 <- tbl %>% select(patient = patient2, infected = patient2_infection)
patient_infection <- bind_rows(infect1, infect2) %>%
  group_by(patient) %>%
  summarize(infected = case_when(
    any(infected == TRUE, na.rm = TRUE) ~ TRUE,
    all(is.na(infected)) ~ NA,            # remain NA if we have no info
    TRUE ~ FALSE
  )) %>%
  ungroup() %>%
  mutate(patient = as.character(patient))

# Ensure all patients present
patient_infection <- tibble(patient = patients) %>%
  left_join(patient_infection, by = "patient") %>%
  mutate(infected = replace_na(infected, FALSE))

# 4) Create an ordered factor for plotting (no labels shown, order arbitrary or by infection)
# We'll order by patient id (alphanumeric). You can change to any ordering.
#tbl_plot <- tbl_sym %>%

# Reorder patients by hierarchical clustering on popANI
popani_mat <- tbl_sym %>%
  select(patient1, patient2, popANI) %>%
  pivot_wider(names_from = patient2, values_from = popANI) %>%
  column_to_rownames("patient1") %>%
  as.matrix()

# Convert to distance (1 - popANI/100) for clustering (higher popANI = closer)
popani_dist <- as.dist(1 - popani_mat/100)
clust <- hclust(popani_dist, method = "average")
patient_order <- clust$labels[clust$order]

# Max popANI
max_popani <- log10(100 -
  instrain_compare_max_popANI %>% 
  filter(popANI !=100 ) %>% pull(popANI) %>% max()
)

#
tbl_plot <- tbl_sym %>%
  mutate(
    patient1 = factor(patient1, levels = patient_order),
    patient2 = factor(patient2, levels = patient_order),
    popANI = if_else(popANI == 100,
                      max_popani * -1,
                      log10(100-popANI) * -1
    )
  )

# Custom popANI color scale: 5 colors, white at popANI=3, linear scale 1-5
popani_breaks <- c(0, 1, 2, 3, 4, 5)
popani_labels <- as.character(popani_breaks)
popani_colors <- c("#2166AC", "#67A9CF", "white", "#F6E8C3", "#F4A582", "#D6604D")

gg <- ggplot(tbl_plot, aes(x = patient1, y = patient2, fill = popANI)) +
  geom_tile(color = NA) +
  scale_fill_gradientn(
    colors = popani_colors,
    values = scales::rescale(popani_breaks),
    breaks = popani_breaks,
    labels = popani_labels,
    limits = c(1, 5),
    oob = scales::squish,
    name = "popANI",
    guide = guide_colorbar(
      title.position = "top",
      title.hjust = 0.5,
      barwidth = unit(1, "cm"),
      barheight = unit(7, "cm")
    )
  ) +
  coord_fixed(clip = "off") +
  theme_minimal(base_size = 10) +
  theme(
    axis.text = element_blank(),
    axis.ticks = element_blank(),
    panel.grid = element_blank(),
    legend.position = "right"
  ) +
  labs(x = NULL, y = NULL) +
  ggnewscale::new_scale_fill() +
  geom_point(
    data = tbl_plot %>% filter(patient1 == levels(tbl_plot$patient1)[1]),
    aes(
      x = -0.7,
      y = patient2,
      fill = patient2_infection
    ),
    shape = 21, color = "black", size = 3, stroke = 0.7, show.legend = FALSE
  ) +
  scale_fill_manual(
    values = c("TRUE" = "#0D7E2B", "FALSE" = "gray75"),
    na.value = "gray50"
  ) +
  ggnewscale::new_scale_fill() +
  geom_point(
    data = tbl_plot %>% filter(patient2 == levels(tbl_plot$patient2)[1]),
    aes(
      x = patient1,
      y = -0.7,
      fill = patient1_infection
    ),
    shape = 21, color = "black", size = 3, stroke = 0.7, show.legend = FALSE
  ) +
  scale_fill_manual(
    values = c("TRUE" = "#0D7E2B", "FALSE" = "gray75"),
    na.value = "gray50"
  )

print(gg)
```

```{r}

```

These results indicate that there are only 10 pairs of patients of patients that share the same strain of *Enterococcus faecium* (i.e., popANI >99.999). I examined these 10 pairs:

```{r}
"sharing_summary <-instrain_compare_max_popANI %>% filter(popANI >= 99.999) %>%
  select(sample1, sample2, patient1, patient2, popANI, conANI, time_between_sample_collection)
```

And it turns out that the 10 pairs correspond to 5 patients which all share the same strain:

```{r}
sharing_matrix <- instrain_compare_max_popANI %>% 
  filter(popANI >= 99.999) %>%
  select(patient1, patient2, popANI) %>%
  pivot_wider(names_from = patient2, values_from = popANI, values_fill = NA, id_cols = patient1) %>%
  mutate(across(everything(), ~ coalesce(.x, NA))) %>%
  ungroup()
```